### DISCUSSION TITLE
Which WASM runtime does it work with?

### DISCUSSION THREAD

**Discussion by kulak** [leptos-rs/leptos#606] [General]
I tried to deploy basic application based on `letpos-tutorial` in docs.

### Build

```
trunk build --all-features
cd dist
wasmedge leptos-tutorial-b66b70904785527b_bg.wasm
```

### Run on `wasmedge`

```
cd dist
wasmedge leptos-tutorial-b66b70904785527b_bg.wasm
```

Alternative build produces the same result with `trunk build`.

Output:

```
[2023-03-01 21:28:07.964] [error] instantiation failed: unknown import, Code: 0x62
[2023-03-01 21:28:07.964] [error]     When linking module: "wbg" , function name: "__wbindgen_object_clone_ref"
[2023-03-01 21:28:07.964] [error]     At AST node: import description
[2023-03-01 21:28:07.964] [error]     At AST node: import section
[2023-03-01 21:28:07.964] [error]     At AST node: module
```

### Run on `wasmtime`

```
wasmtime leptos-tutorial-b66b70904785527b_bg.wasm
Error: failed to run main module `leptos-tutorial-b66b70904785527b_bg.wasm`

Caused by:
    0: failed to instantiate "leptos-tutorial-b66b70904785527b_bg.wasm"
    1: unknown import: `wbg::__wbindgen_object_clone_ref` has not been defined
```

### Works

`trunk serve --open` works just fine.

**Comment by kulak** [comment]
I think I am completely wrong here.  At least the tutorial part is meant to be executed by the browser.

**Comment by gbj** [comment]
Right, everything in the tutorial so far is for client-side rendering in the browser.

You can do server-side rendering with native binaries with Leptos very easily.

You can also do server-side rendering in WASM runtimes like run within V8 (Cloudflare Workers, Deno, etc.)

You can also do server-side rendering in `wasm32-wasi` environments, depending on the environment, with some additional configuration that has to do with the interaction between `wasm-bindgen` and `wasm32-wasi` targets that don't exist in a JavaScript world. For example, in `wasmtime` you can add a `--trap-unknown-imports` flag that prevents the panic you got. I've never seen `wasmedge` before so I don't know about that.
